app/ante/msg_exec.go

```diff
@@ -0,0 +1,46 @@
+package ante
+
+import (
+	blobtypes "github.com/celestiaorg/celestia-app/v3/x/blob/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+)
+
+var _ sdk.AnteDecorator = MsgExecDecorator{}
+
+// MsgExecDecorator ensures that the tx does not contain a MsgExec with a
+// nested MsgExec or MsgPayForBlobs.
+type MsgExecDecorator struct{}
+
+func NewMsgExecDecorator() *MsgExecDecorator {
+	return &MsgExecDecorator{}
+}
+
+func (mgk MsgExecDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	// Only apply this decorator in checkTx.
+	// TODO: in v4, consider enforcing this in ProcessProposal too.
+	if !ctx.IsCheckTx() {
+		return next(ctx, tx, simulate)
+	}
```

Comment:
Do we want to remove this for this PR to `main` and only add it in the backport to v3.x?

Reply:
Yea can do. I'll split out a new PR that targets v3.x which includes this and remove it from this PR.